src/NuGet.Core/NuGet.ProjectModel/TargetFrameworkInformation.cs

```diff
@@ -39,23 +38,57 @@ public class TargetFrameworkInformation : IEquatable<TargetFrameworkInformation>
         /// <summary>
         /// List of dependencies that are not part of the graph resolution.
         /// </summary>
-        public IList<DownloadDependency> DownloadDependencies { get; } = new List<DownloadDependency>();
+        public IList<DownloadDependency> DownloadDependencies { get; }
 
         /// <summary>
         /// List of the package versions defined in the Central package versions management file. 
         /// </summary>
-        public IDictionary<string, CentralPackageVersion> CentralPackageVersions { get; } = new Dictionary<string, CentralPackageVersion>(StringComparer.OrdinalIgnoreCase);
+        public IDictionary<string, CentralPackageVersion> CentralPackageVersions { get; }
 
         /// <summary>
         /// A set of unique FrameworkReferences
         /// </summary>
-        public ISet<FrameworkDependency> FrameworkReferences { get; } = new HashSet<FrameworkDependency>();
+        public ISet<FrameworkDependency> FrameworkReferences { get; }
 
         /// <summary>
         /// The project provided runtime.json
         /// </summary>
         public string RuntimeIdentifierGraphPath { get; set; }
 
+        public TargetFrameworkInformation()
+        {
+            TargetAlias = string.Empty;
+            Dependencies = new List<LibraryDependency>();
+            Imports = new List<NuGetFramework>();
+            DownloadDependencies = new List<DownloadDependency>();
+            CentralPackageVersions = new Dictionary<string, CentralPackageVersion>(StringComparer.OrdinalIgnoreCase);
+            FrameworkReferences = new HashSet<FrameworkDependency>();
+        }
+
+        internal TargetFrameworkInformation(TargetFrameworkInformation cloneFrom)
+        {
+            TargetAlias = cloneFrom.TargetAlias;
+            FrameworkName = cloneFrom.FrameworkName;
+            Dependencies = CloneList(cloneFrom.Dependencies, static dep => dep.Clone());
+            Imports = new List<NuGetFramework>(cloneFrom.Imports);
+            AssetTargetFallback = cloneFrom.AssetTargetFallback;
+            Warn = cloneFrom.Warn;
+            DownloadDependencies = CloneList(cloneFrom.DownloadDependencies, static dep => dep.Clone());
+            CentralPackageVersions = new Dictionary<string, CentralPackageVersion>(cloneFrom.CentralPackageVersions, StringComparer.OrdinalIgnoreCase);
+            FrameworkReferences = new HashSet<FrameworkDependency>(cloneFrom.FrameworkReferences);
```

Comment:
I think we can use [`HashSet<T>(Int32, IEqualityComparer<T>)`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.-ctor?view=net-7.0#system-collections-generic-hashset-1-ctor(system-int32-system-collections-generic-iequalitycomparer((-0)))) constructor because we know the size of the collection.

Reply:
No such constructor exists, in .NET Framework at least. In this case however, the `HashSet<T>` constructor sniffs the `IEnumerable<T>` to see whether it's a `HashSet<T>`. It will be, and will then optimise the copy. So I think we're good here.

![image](https://github.com/NuGet/NuGet.Client/assets/350947/065f99b0-19be-448d-a077-c76eac55beb7)

The comparers will be equal.

Reply:
Sorry for the confusion. I overlooked the constructor definition. I thought the second argument of this constructor is of type IEnumerable but I was wrong it is IEqualityComparerInstead [HashSet<T>(Int32, IEqualityComparer<T>)](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.-ctor?view=net-7.0#system-collections-generic-hashset-1-ctor(system-int32-system-collections-generic-iequalitycomparer((-0)))).